### ISSUE TITLE
Popup opens by itself on office.com

### ISSUE THREAD

**Issue by Illiou** [degecko/prevue#6]
I just noticed that when I go to [office.com](office.com) while being logged in, it automatically opens the extension's sidebar and displays an error saying it couldn't log me in because JavaScript needs to be enabled (doesn't affect my actual login state though). If I disable the extention nothing like that shows up.

How is it possible that the popup opens by itself? Can you replicate this?

![Screenshot 2022-10-01 at 21-24-17 Microsoft Office Home](https://user-images.githubusercontent.com/28632830/193425355-64729779-ed93-424e-ad8d-37e28dddb4e5.png)

**Comment by degecko** [comment]
Unfortunately, I cannot replicate this. I logged in, but nothing happens.

Please check to see if you use a trigger which activates the extension when pressing a button and moving the mouse over a link. I'm thinking that you're working on something, maybe press ctrl + c/v to copy/paste and you're also moving the cursor over something at the same time.

You could simply disable all triggers and check if it works again also. Then re-enable the triggers one by one until you find out what's causing it.

**Comment by Illiou** [comment]
I had only activated the "Cursor over" Trigger but interestingly even after disabling that, i.e. with all triggers disabled, the popup still opens. All I do is enter the url into the address bar and hit enter. After a few seconds it appears. In fact, no matter how I change the settings, as long as the extention is enabled the popup is opened. It also seems to ignore some settings, for example even though I have set the popup to always open on the right, only this one opens on the left side. (For the original screenshot I had used the button to switch sides, as an easy way to avoid the document names being in the screenshot.)

Is there a way for me to debug how the popup is triggered?

**Comment by degecko** [comment]
> Is there a way for me to debug how the popup is triggered?

I can't think of any specific thing.

I would try it in different contexts to see if it still triggers itself:

- With other extensions disabled;
- In incognito mode;
- On Firefox ([Prevue for Firefox](https://addons.mozilla.org/en-US/firefox/addon/prevue-popup/))

Also, if you've tried to run it yourself from a local version, maybe you now have two extensions enabled?

**Comment by Illiou** [comment]
Ah I'm sorry, I should have mentioned before that I'm on Firefox. Newest version on Windows 10.

I can actually reproduce it on a fresh portable Firefox. Just started it, installed the extension and then logged into office.com.
Maybe you can try to reproduce it again on Firefox too.

**Comment by degecko** [comment]
@Illiou All good, I was able to reproduce and fix the issue in Firefox. Thanks!

I've implemented the fix in version 0.4.7, which should be installed automatically in a few hours, or manually, if you go to: about:addons > Click the gear icon and select "Check for Updates".

Should not open by itself anymore.

**Comment by Illiou** [comment]
Great, thanks! Can confirm this fixes it.

I understand you now use a suffix on the url to prevent unintended opening, but I'm curious how a website can interact like this with the extension in the first place? What caused it to open despite all triggers being disabled?

**Comment by degecko** [comment]
> I understand you now use a suffix on the url to prevent unintended opening, but I'm curious how a website can interact like this with the extension in the first place? What caused it to open despite all triggers being disabled?

It was opening for a different reason. Some URLs, such as the ones from Twitter posts, are being proxied through their own t.co URL shortening URL service. If I were to preview the URLs like that, it would result in disabling the CSP and X-Frame-Options for that short URL, and after redirecting to the final URL, the preview would fail, because the CSP and X-Frame-Options headers would still be there on the final URL.

I needed a way for the extension to find the last URL in the redirect chain, to be able to remove the CSP and X-Frame-Options headers from that one, to be able to open it in the iframe of the extension. So, to do that, i'm injecting in the opening iframe some code that reports back to the extension the current URL of the iframe, which, after loading all redirects, will be the final URL.

That piece of reporting mechanism was implemented very primitively in Firefox, because they're using older versions the extension APIs, which aren't allowing me to tell if the iframe belongs to Prevue or not. In Chrome, I have access to check if the opening iframe belongs to the extension, but in Firefox I had to inject that code in all iframes on all pages.

So, basically, the code that reported the final URL was injected in all iframes, which, in office.com's case, included one of their own iframes. That URL was reported to Prevue, which in turn opened the popup for the reported final URL, because it's hard to tell who initiated the "report URL" request.

I could've fixed that by checking if a Prevue popup was initiated first. I've tried that, but I found something which still didn't work right, so I had to change the logic again.

Therefore, I added the #_prevue thing in the previewing URL, to be able to check that when I'm asking to report back the final URL.

The extension is full of hacks like this, because the extension APIs are extremely restrictive when working with iframes on websites, both on Firefox and Chrome.
